packages/vite/src/node/plugins/dynamicImportVars.ts

```diff
@@ -0,0 +1,214 @@
+  rawPattern: string
+}
+
+const dynamicImportHelper = (glob: Record<string, any>, path: string) => {
+  const v = glob[path]
+  if (v) {
+    return typeof v === 'function' ? v() : Promise.resolve(v)
+  }
+  return new Promise((_, reject) => {
+    ;(typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
+      reject.bind(null, new Error('Unknown variable dynamic import: ' + path))
+    )
+  })
+}
+
+function parseDynamicImportPattern(
+  strings: string
+): DynamicImportPattern | null {
+  const filename = strings.slice(1, -1)
+  const rawQuery = parseRequest(filename)
+  let globParams: DynamicImportRequest | null = null
+  const ast = (
+    parseJS(strings, {
+      ecmaVersion: 'latest',
+      sourceType: 'module'
+    }) as any
+  ).body[0].expression
+
+  const userPatternQuery = dynamicImportToGlob(ast, filename)
+  if (!userPatternQuery) {
+    return null
+  }
+
+  const [userPattern] = userPatternQuery.split('?', 2)
+  const [rawPattern] = filename.split('?', 2)
+
+  if (rawQuery?.raw !== undefined) {
+    globParams = { as: 'raw' }
+  }
+
+  return {
+    globParams,
+    userPattern,
+    rawPattern
+  }
+}
+
+export async function transformDynamicImport(
+  importSource: string,
+  importer: string,
+  resolve: (
+    url: string,
+    importer?: string
+  ) => Promise<string | undefined> | string | undefined
+): Promise<{
+  glob: string
+  pattern: string
+  rawPattern: string
+} | null> {
+  if (importSource[1] !== '.' && importSource[1] !== '/') {
+    const resolvedFileName = await resolve(importSource.slice(1, -1), importer)
+    if (!resolvedFileName) {
+      return null
+    }
+    const relativeFileName = posix.relative(
+      posix.dirname(normalizePath(importer)),
+      normalizePath(resolvedFileName)
+    )
+    importSource = normalizePath(
+      '`' + (relativeFileName[0] === '.' ? '' : './') + relativeFileName + '`'
+    )
+  }
+
+  const dynamicImportPattern = parseDynamicImportPattern(importSource)
+  if (!dynamicImportPattern) {
+    return null
+  }
+  const { globParams, rawPattern, userPattern } = dynamicImportPattern
+  const params = globParams
+    ? `, ${JSON.stringify({ ...globParams, import: '*' })}`
+    : ''
+  const exp = `(import.meta.glob(${JSON.stringify(userPattern)}${params}))`
+
+  return {
+    rawPattern,
+    pattern: userPattern,
+    glob: exp
+  }
+}
+
+export function dynamicImportVarsPlugin(config: ResolvedConfig): Plugin {
+  const resolve = config.createResolver({
+    preferRelative: true,
+    tryIndex: false,
+    extensions: []
+  })
+  const { include, exclude, warnOnError } =
+    config.build.dynamicImportVarsOptions
+  const filter = createFilter(include, exclude)
+
+  return {
+    name: 'vite:dynamic-import-vars',
+
+    resolveId(id) {
+      if (id === dynamicImportHelperId) {
+        return id
+      }
+    },
+
+    load(id) {
+      if (id === dynamicImportHelperId) {
+        return 'export default' + dynamicImportHelper.toString()
```

Comment:
```suggestion
        return 'export default ' + dynamicImportHelper.toString()
```

It might be safer if we leave a space for the inlined function.